ompi/mca/osc/ucx/osc_ucx_passive_target.c

```diff
@@ -0,0 +1,374 @@
+        }
+        assert(opal_list_is_empty(&module->outstanding_locks) == true);
+    } else {
+        assert(module->epoch_type.access != PASSIVE_EPOCH);
+        OPAL_LIST_FOREACH(item, &module->outstanding_locks, ompi_osc_ucx_lock_t) {
+            if (item->target_rank == target) {
+                return OMPI_ERR_RMA_SYNC;
+            }
+        }
+    }
+
+    module->epoch_type.access = PASSIVE_EPOCH;
+    module->lock_count++;
+    assert(module->lock_count <= ompi_comm_size(module->comm));
+
+    lock = OBJ_NEW(ompi_osc_ucx_lock_t);
+    lock->target_rank = target;
+
+    if (lock_type == MPI_LOCK_EXCLUSIVE) {
+        lock->type = LOCK_EXCLUSIVE;
+    } else {
+        lock->type = LOCK_SHARED;
+    }
+
+    if ((assert & MPI_MODE_NOCHECK) == 0) {
+        lock->is_nocheck = false;
+        if (lock_type == MPI_LOCK_EXCLUSIVE) {
+            ret = start_exclusive(module, target);
+        } else {
+            ret = start_shared(module, target);
+        }
+    } else {
+        lock->is_nocheck = true;
+    }
+
+    if (ret == OMPI_SUCCESS) {
+        opal_list_append(&module->outstanding_locks, &lock->super);
+    } else {
+        OBJ_RELEASE(lock);
+    }
+
+    return ret;
+}
+
+int ompi_osc_ucx_unlock(int target, struct ompi_win_t *win) {
+    ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t *)win->w_osc_module;
+    ompi_osc_ucx_lock_t *lock = NULL, *item = NULL;
+    ucs_status_t status;
+    int ret = OMPI_SUCCESS;
+    ucp_ep_h ep;
+
+    if (module->epoch_type.access != PASSIVE_EPOCH) {
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    OPAL_LIST_FOREACH(item, &module->outstanding_locks, ompi_osc_ucx_lock_t) {
+        if (item->target_rank == target) {
+            lock = item;
+            break;
+        }
+    }
+
+    if (lock == NULL) {
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    opal_list_remove_item(&module->outstanding_locks, &lock->super);
+
+    ep = OSC_UCX_GET_EP(module->comm, target);
+    status = ucp_ep_flush(ep);
+    if (status != UCS_OK) {
+        return OMPI_ERROR;
+    }
+
+    module->global_ops_num -= module->per_target_ops_nums[target];
+    module->per_target_ops_nums[target] = 0;
+
+    if (lock->is_nocheck == false) {
+        if (lock->type == LOCK_EXCLUSIVE) {
+            ret = end_exclusive(module, target);
+        } else {
+            ret = end_shared(module, target);
+        }
+    }
+
+    OBJ_RELEASE(lock);
+
+    module->lock_count--;
+    assert(module->lock_count >= 0);
+    if (module->lock_count == 0) {
+        module->epoch_type.access = NONE_EPOCH;
+        assert(module->global_ops_num == 0);
+    }
+
+    return ret;
+}
+
+int ompi_osc_ucx_lock_all(int assert, struct ompi_win_t *win) {
+    ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t*) win->w_osc_module;
+    ompi_osc_ucx_lock_t* lock;
+    int ret = OMPI_SUCCESS;
+
+    if (module->epoch_type.access != NONE_EPOCH &&
+        module->epoch_type.access != FENCE_EPOCH) {
```

Comment:
(now L236)
I think this is the same problem with the first long discussion. When the user call Fence with no assert, MPI library cannot decide if epoch is closed or not, and the state should always be FENCE_EPOCH.